### PR TITLE
Improve editor camera

### LINKED ISSUE #165: Editor Camera Improvements
The current editor camera is not nice. There are some improvements needed
- [x] Camera should move with MMB on xz Plane
- [x] Camera should move on local xz Plane on Alt+MMB
- [x] Camera should orbit around intersection point in view direction with Shift+MMB
- [x] Camera should zoom on Ctrl+MMB
- [x] Camera should be moveable up and down
- [x] Camera should zoom on scroll
- [x] Cursor should end where it was captured

Some of those improvements should be ported to the game camera as well.

### PR DESCRIPTION
**Pull request by HaSa1002**
This PR improves the editor camera a lot.

When pressing `RMB`, you can fly the camera like in a FP game. Use `W` (forward), `A` (left), `S` (back), `D` (right), `E` (up), `Q` (down) and `Shift` (quick).
When pressing `MMB`, you can pan the camera on the xz-Plane.
When pressing `Shift+MMB`, you can rotate the camera around the ground (should be refactored in the future to enable objects) or around a point in the distance of the `zoom_level`.
When pressing `Ctrl+MMB` or using the scroll wheel, you can zoom. Please note, it is rather a movement on the forward axis rather than a real zoom.
When pressing `Alt+MMB`, you can pan the camera on its local xy-Plane.

Fixes #165

### CODE REVIEW

**Comment by HaSa1002** [comment]
Split into camera base and editor camera to be able to implement a seated and fly camera for the game with the same functions. I started doing that, but it is a can of worms, because the camera is *partly* controlled by the player script and I need the menu refactoring, because the old menus eat input everywhere. Once that is done, the old FreeCamera can be deleted.

TL;DR:
It is future-proofing, but only because I don't want to bloat this PR and it takes no time to de-future-proof this feature into the desired use case

**Comment by HaSa1002** [comment]
rebased

File: src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	zoom_level = clamp(zoom_level - offset, 0.2, 250)
+	translate(Vector3(0,0, prev_zoom_level - zoom_level))
+
+
+func _rotate_local(offset: Vector2) -> void:
+		rot -= offset
+		rot.y = clamp(rot.y, -0.5*PI, 0.5*PI)
+		transform.basis = Basis()
+		rotate_object_local(Vector3.UP, rot.x)
+		rotate_object_local(Vector3.RIGHT, rot.y)
```
Comment (DasCapschen) [inline src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd]: sure you don't want to switch x and y? You know, since Y is the UP axis, and X the right axis...

File: src/addons/Libre_Train_Sim_Editor/Data/Scripts/editor_camera.gd [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+export var pan_move_sensitivity: float = 0.25
+
+
+var velocity := Vector3.ZERO
+var is_moving_first_person := false
+var is_panning := false
+
+
+# TODO: F to focus on selected object
+# TODO: Fix Pan speed depending on distance to ground or zoom
```
Comment (DasCapschen) [inline src/addons/Libre_Train_Sim_Editor/Data/Scripts/editor_camera.gd:17]: want to implement this, or leave it as TODO?

**Review by DasCapschen — COMMENTED** [review]
looks good, check comments

File: src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd (hunk 1 above)
Comment (HaSa1002) [reply src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd]: Yep. Try it 😅

File: src/addons/Libre_Train_Sim_Editor/Data/Scripts/editor_camera.gd (hunk 2 above)
Comment (HaSa1002) [reply src/addons/Libre_Train_Sim_Editor/Data/Scripts/editor_camera.gd:17]: first one will stay there as TODO. Second... idk the zoom is not really a zoom, because I would need to add object selection which I didn't want to do for now. The objects would need collisions which they don't have now, iirc

Comment (HaSa1002) [reply src/addons/Libre_Train_Sim_Editor/Data/Scripts/editor_camera.gd:17]: fixed

File: src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd (hunk 1 above)
Comment (HaSa1002) [reply src/addons/Libre_Train_Sim_Editor/Data/Scripts/camera_base.gd]: fixed with comment

**Comment by HaSa1002** [comment]
there you go :)

**Review by DasCapschen — APPROVED** [review]
lgtm
